src/tools/ceph_dedup_tool.cc

```diff
@@ -502,6 +514,564 @@ void ChunkScrub::chunk_scrub_common()
+    ObjectCursor current,
+    ObjectCursor end,
+    size_t max_object_count);
+  std::set<size_t> sample_object(size_t count);
+  void try_object_dedup_and_accumulate_result(ObjectItem& object);
+  bool ok_to_dedup_all();
+  void flush_duplicable_object(ObjectItem& object);
+  AioCompletion* set_chunk_duplicated(chunk_t& chunk);
+  void mark_non_dedup(ObjectCursor start, ObjectCursor end);
+  bufferlist read_object(ObjectItem& object);
+  std::vector<std::tuple<bufferlist, pair<uint64_t, uint64_t>>> do_cdc(
+    ObjectItem& object,
+    bufferlist& data);
+  std::string generate_fingerprint(bufferlist chunk_data);
+  bool check_whole_object_dedupable(size_t dedup_size, size_t total_size);
+  bool is_dirty(ObjectItem& object);
+  AioCompletion* do_async_evict(string oid);
+
+  Rados rados;
+  IoCtx chunk_io_ctx;
+  crawl_mode_t mode;
+  uint32_t sampling_ratio;
+  std::list<chunk_t> duplicable_chunks;
+  size_t total_duplicated_size = 0;
+  size_t total_object_size = 0;
+  size_t chunk_dedup_threshold;
+  size_t object_dedup_threshold;
+
+  class FpStore {
+    public:
+      struct fp_store_entry_t {
+        size_t duplication_count = 1;
+        std::list<chunk_t> found_chunks;
+        bool processed = false;
+      };
+
+      bool find(string& fp) {
+        std::shared_lock lock(fingerprint_lock);
+        auto found_item = fp_map.find(fp);
+        if (found_item != fp_map.end()) {
+          return true;
+        }
+        return false;
+      }
+
+      void add(chunk_t& chunk, std::list<chunk_t>& duplicable_chunks) {
+        std::unique_lock lock(fingerprint_lock);
+        auto found_iter = fp_map.find(chunk.fingerprint);
+        if (found_iter != fp_map.end()) {
+          auto& target = found_iter->second;
+          target.duplication_count++;
+          target.found_chunks.push_back(chunk);
+          if (target.duplication_count >= dedup_threshold) {
+            if (target.processed == false) {
+              target.processed = true;
+              // When a fingerprint firstly detected to be duplicated more than
+              // threshold, add all previously found chunks to duplicable_chunks
+              duplicable_chunks.splice(duplicable_chunks.begin(), target.found_chunks);
+            } else {
+              duplicable_chunks.push_back(chunk);
+            }
+          }
+        } else {
+          fp_store_entry_t fp_entry;
+          fp_entry.found_chunks.push_back(chunk);
+          fp_map.insert({chunk.fingerprint, fp_entry});
+        }
+      }
+
+      void init(size_t dedup_threshold_) {
+        std::unique_lock lock(fingerprint_lock);
+        fp_map.clear();
+        dedup_threshold = dedup_threshold_;
+      }
+
+    private:
+      size_t dedup_threshold = -1;
+      std::unordered_map<std::string, fp_store_entry_t> fp_map; 
+      std::shared_mutex fingerprint_lock;
+  };
+
+  std::set<std::string> oid_for_evict;
+  static FpStore fp_store;
+  static std::unordered_set<std::string> flushed_objects;
```

Comment:
I don't actually understand what this is for.  Each worker thread will only consider an object once and the set of objects considered by different workers is disjoint.  Why do we need to remember this?